### PR TITLE
[Feature] DFlash + ACLGraph FDO/FULL mode support

### PR DESCRIPTION
**Pull request by Liuchenbing-2026**
## Summary

Enable DFlash speculative decoding to work correctly with ACLGraph compilation modes (FULL_DECODE_ONLY / FULL) on Ascend NPU.

DFlash is a block-level diffusion-based speculative decoding method ([vllm PR #36847](https://github.com/vllm-project/vllm/pull/36847)) that generates k draft tokens in parallel. Combined with FULL_DECODE_ONLY graph mode, it achieves **5.8x** throughput speedup (175 tok/s vs 30 tok/s baseline) on Qwen3-8B.

### Changes

1. **CompilationTimes + eagle inputs fallback** (`worker.py`, `eagle_proposer.py`)
   - Wrap `compile_or_warm_up_model` return value as `CompilationTimes` namedtuple
   - Add pure Python fallback for `npu_copy_and_expand_eagle_inputs`

2. **update_stream initialization** (`draft_proposer.py`)
   - Initialize `update_stream` + `ACLGraphWrapper` in `_maybe_share_lm_head` for FULL_DECODE_ONLY/FULL mode

3. **DFlash + ACLGraph core fixes** (`attention_v1.py`, `dflash_proposer.py`, `eagle_proposer.py`)
   - Dynamic `sparse_mode` for non-causal cross-attention (fixes acceptance 0.62% -> 30%)
   - Build DFlash-specific non-causal `attn_metadata` for FULL graph capture
   - Move `prepare_context_kv` outside graph boundary (variable-length input incompatible with graph replay)
   - Sync DFlash `query_start_loc` to runner buffer before FIA padding
   - Fix TND dimension constraint (`actual_seq_lengths_q` vs `num_tokens`)
   - Clamp garbage token IDs after first graph capture (prevents NPU error 507057)
   - Clear `draft_graph_params` after `dummy_run` warmup

4. **hidden_states guard** (`eagle_proposer.py`)
   - Guard `hidden_states` copy with `pass_hidden_states_to_model` check

5. **ACLGraphWrapper lazy capture fix** (`acl_graph.py`)
   - Replace `validate_cudagraph_capturing_enabled()` with `set_cudagraph_capturing_enabled(True)` for draft model's lazy capture path
   - Use `defaultdict(list)` for `GraphParams` to prevent `KeyError`

### Performance (Qwen3-8B, MATH-500, 50 prompts, concurrency=1)

| Mode | Throughput | TTFT | TPOT | Acceptance | Speedup |
|------|-----------|------|------|-----------|---------|
| No spec (baseline) | 30.33 tok/s | 119.23 ms | 32.62 ms | - | 1.0x |
| DFlash + PIECEWISE (spec=15) | 142.23 tok/s | 100.33 ms | 6.75 ms | 38.30% | 4.7x |
| **DFlash + FULL_DECODE_ONLY (spec=15)** | **175.23 tok/s** | 96.35 ms | **5.39 ms** | 29.99% | **5.8x** |
| DFlash + FULL (spec=15) | 171.61 tok/s | **45.40 ms** | 5.72 ms | 27.62% | 5.7x |
| Eagle3 + PIECEWISE (spec=3) | 45.30 tok/s | 72.00 ms | 21.91 ms | 36.43% | 1.5x |
| Eagle3 + FULL_DECODE_ONLY (spec=3) | 74.40 tok/s | 98.46 ms | 13.13 ms | 36.66% | 2.5x |
| Eagle3 + FULL (spec=3) | 74.80 tok/s | 52.46 ms | 13.25 ms | 36.50% | 2.5x |

### Compatibility

Verified no regression on non-speculative serving, Eagle3, and EAGLE draft_model:

| Scenario | Model | Method | Mode | Result |
|----------|-------|--------|------|--------|
| No spec (baseline) | Qwen3-8B | - | PIECEWISE | Pass |
| Eagle3 | Qwen3-8B | eagle3, spec=3 | FULL | Pass |
| EAGLE draft_model | LLaMA-3.1-8B | draft_model, spec=8 | FULL | Pass |
| DFlash | Qwen3-8B | dflash, spec=15 | FULL_DECODE_ONLY | Pass |
| DFlash | Qwen3-8B | dflash, spec=15 | FULL | Pass |

## Test plan

- [x] DFlash + Qwen3-8B with PIECEWISE / FULL_DECODE_ONLY / FULL modes
- [x] Eagle3 + Qwen3-8B with FULL mode
- [x] EAGLE draft_model + LLaMA-8B with FULL mode
- [x] Non-speculative baseline serving

### CODE REVIEW

File: vllm_ascend/spec_decode/dflash_proposer.py [hunk 1]
```diff
@@ -166,35 +166,61 @@ def dummy_run(
...
+        if self.use_cuda_graph and not is_profile and self.block_table_tensor_clone is None:
+            self.block_table_tensor_clone = torch.zeros(
+                (
+                    self.runner.max_num_tokens + 2 * self.pcp_size * self.runner.max_num_reqs,
+                    self.runner.input_batch.block_table[0].get_device_tensor().shape[1],
+                ),
+                dtype=torch.int32,
+                device=self.device,
+                pin_memory=self.runner.pin_memory,
+            )
```
Comment (slippersss) [inline vllm_ascend/spec_decode/dflash_proposer.py:180]: As far as I know, this piece of code is for eagle+pcp+full, and I think it should not appear here. If you are just trying to use it to solve ci break, it is not necessary since it has been solved now.

File: vllm_ascend/compilation/acl_graph.py [hunk 2]
```diff
@@ -129,8 +130,8 @@ def __call__(self, *args, **kwargs):
                 # shape. E.g. we only log it for the first subgraph in
                 # piecewise mode.
                 logger.debug("Capturing a aclgraph on (%s,%s)", self.runtime_mode.name, entry.batch_descriptor)
-            # validate that aclgraph capturing is legal at this point.
-            validate_cudagraph_capturing_enabled()
+            # enable aclgraph capturing for lazy capture path (draft model)
+            set_cudagraph_capturing_enabled(True)
```
Comment (slippersss) [inline vllm_ascend/compilation/acl_graph.py:139]: This is weird. I do not think it is reasonable. All capturing should be done at the very beginning of the running service, in my understanding, `cudagraph_capturing_enabled` will be true during this period. Except that you are running into here when you are actually inferencing requests, and that is unexpected.

File: vllm_ascend/spec_decode/dflash_proposer.py [hunk 3]
```diff
@@ -219,23 +245,54 @@ def dummy_run(
...
+                    draft_gp.handles[key] = []
+                for key in list(draft_gp.workspaces.keys()):
+                    draft_gp.workspaces[key] = None
+            # Invalidate ACLGraphWrapper's captured graph entries so the first
+            # real _propose() call re-captures with correct runtime params.
+            # During dummy_run, self.model (ACLGraphWrapper) captured a graph
+            # with stale dummy values. Without clearing, the first real call
+            # would replay that stale graph instead of re-capturing.
+            if isinstance(self.model, ACLGraphWrapper):
+                self.model.concrete_aclgraph_entries.clear()
```
Comment (slippersss) [inline vllm_ascend/spec_decode/dflash_proposer.py:277]: Pretty strange. Reset graph params at the end of `dummy_run` and try to re-capture in `_propose`? I think we should solve the problem why we need to capture when inferencing instead of patching everywhere.

File: vllm_ascend/spec_decode/dflash_proposer.py (hunk 1 above)
Comment (Liuchenbing-2026) [reply vllm_ascend/spec_decode/dflash_proposer.py:180]: I tried removing it and it crashed — `AscendDflashProposer` doesn't override `_propose`, so it goes through `AscendSpecDecodeBaseProposer._propose` (eagle_proposer.py:692) which asserts on `block_table_tensor_clone` in graph mode. It's used to clone the block_table so multi-step inference doesn't overwrite earlier steps.

File: vllm_ascend/compilation/acl_graph.py (hunk 2 above)
Comment (Liuchenbing-2026) [reply vllm_ascend/compilation/acl_graph.py:139]: This line isn't from this PR — it's existing code in `ACLGraphWrapper.__call__`. DFlash hits it because the draft model does non-causal cross-attention, so we can't capture all shapes upfront like Eagle does. It captures on first encounter of each shape, then replays from cache after that.

File: vllm_ascend/spec_decode/eagle_proposer.py [hunk 4]
```diff
@@ -812,6 +822,15 @@ def _propose(
...
             )
 
+            # NPU graph capture does not execute kernels — only records
+            # operations. The first ACLGraphWrapper call (capture) returns
+            # uninitialized garbage in the output tensor. Clamp token IDs
+            # to valid vocabulary range to prevent OOB embedding access
+            # in the target model (which would cause NPU error 507057).
+            vocab_size = self.vllm_config.model_config.get_vocab_size()
+            if draft_token_ids is not None and isinstance(draft_token_ids, torch.Tensor):
+                draft_token_ids.clamp_(0, vocab_size - 1)
```
Comment (lilinsiman) [inline vllm_ascend/spec_decode/eagle_proposer.py:867]: Just for aclgraph or not? If it's just for aclgraph, then add an if branch. If not, please explain why this processing wasn't needed in the propose function before—was it to fix some bug?

**Comment by yiz-liu** [comment]
@FutureSkyFly Please rebase and fix failed test cases. And is this a duplicate of #8906 ?

**Comment by Liuchenbing-2026** [comment]
Closing this PR — it has been superseded by #8906, which contains all commits from this PR plus 6 additional fixes:

- `b58bcc04` fix(gdn): increase MAX_MTP from 8 to 12 for higher spec values
- `1c43f32c` fix(dflash): use kernel_block_size for slot mapping in Triton kernel (NPU 507035 fix)
- `1dfe8e74` fix(dflash): enable aux_hidden_state collection for DFlash on Qwen3.5 (garbled output fix)
- `11b60bf7` fix: align dummy_run mm input handling with upstream
- `85fa3d13` fix: support DFlash with Qwen3.5 hybrid attention architecture
- `cd54c20a` fix: unify DFlash FULL mode graph capture to wrap _runnable (FDO/FULL acceptance rate +33-35%)

Please review #8906 instead.

**Comment by Liuchenbing-2026** [comment]
Superseded by #8906
